perf: don't extract gzipped backups

akhilnarang opened the pull request:
Read a few lines from the gzipped file directly wherever possible for basic validation
Don't create a temporary uncompressed file, directly pipe the output of `gzip -dc` to mariadb
Cleanup backup decryption handling code

Resolves #22546

akhilnarang committed:
perf: don't extract backup files unless required

Read from the gzipped file wherever possible

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/installer.py:
```diff
@@ -777,48 +777,63 @@ def is_downgrade(sql_file_path, verbose=False):
-				all_apps = [x[-3:] for x in apps_list]
+	header = get_db_dump_header(sql_file_path).split("\n")
+	if "Version" in header[0]:
+		backup_version = header[0].split(":")[-1].strip()
```
umm, can we do better than this? Let's not manually parse things and make this future proof/extensible. 

Use `ini` or `TOML` format and wrap it in begin/after lines?

ankush commented:
We also need to move away from extracting it and instead directly pipe to mysql: https://github.com/frappe/frappe/blob/287e13522cbae1f2e3656f9906381b877105b207/frappe/installer.py#L710-L721  

Otherwise, we don't gain much from this change alone as of now.

akhilnarang committed:
fix(Backup): automatically rollback decryption when object is being deleted

This allows us to not have to call it everytime before returning

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat(installer): drop actual gzip file extraction

Use `zgrep` to check for table name match where required

Also use a table that's at the top of the dump files (`__Auth`)

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat(db_manager): avoid extraction of DB dump if gzipped

Use `gzip -cd` to directly get the contents onto stdout and pipe to mariadb

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat(restore): handle encrypted backups better

Determine the mimetype based on the file contents instead of waiting for an exception
Cleaner + no need of duplicate code

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: drop use of `magic` library, run `file` command directly

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(postgres/setup): use `gzip` to get backup contents if the file is an archive

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/utils/backups.py:
```diff
@@ -700,14 +704,16 @@ def decryption_rollback(self):
 				os.remove(self.file_path.rstrip(".gz"))
 			os.rename(self.file_path + ".gpg", self.file_path)
 
+	def __del__(self):
```
overriding `__del__` is usually a bad practice. It's an internal method used by garbage collector, what we are doing is more suited for contextmanager.

akhilnarang replied:
Based on what I understood its mostly safe enough unless the class has cyclic references.

Context managers seem to become a bit messy with the way this code has to be used, will see if I can figure out a clean way to implement

ankush commented on frappe/installer.py:
```diff
@@ -837,43 +829,60 @@ def partial_restore(sql_file_path, verbose=False):
 
-	if not os.path.getsize(path):
+	if path.endswith(".gz"):
+		executable_name = "zgrep"
```
Is this widely available? Check frappe docker containers and typical ubuntu distros. If not better to write a stream using GZIP + grep.

akhilnarang replied:
Yep, was present by default on Ubuntu, Debian, and `frappe/erpnext` image

ankush commented on frappe/installer.py:
```diff
@@ -777,53 +738,84 @@ def is_downgrade(sql_file_path, verbose=False):
+	backup_version = None
+	try:
+		backup_version = extract_version_from_dump(sql_file_path)
+	except Exception:
```
Exceptions should not be used for control flow if possible, if the return value is None that indicates better parsing failure? :thinking:

akhilnarang replied:
Yep will adjust it a bit, thought of this so that the old logic could be a simple fallback incase something goes wrong, but the logic inside the function is relatively "safe" so can be avoided.

ankush commented on frappe/installer.py:
```diff
@@ -707,20 +682,6 @@ def convert_archive_content(sql_file_path):
 		old_sql_file_path.unlink()
 
 
-def extract_sql_gzip(sql_gz_path):
```
Just check if it's not used anywhere. Sourcegraph is a nice tool for this. 

https://sourcegraph.com/search?q=context:global+extract_sql_gzip+repo:%5Egithub%5C.com/frappe/.*&patternType=standard&sm=0&groupBy=repo

akhilnarang replied:
Ah cool, thanks.

ankush reviewed (changes requested):
Add tests for all known cases please. 

E.g. backups taken before this PR are restorable. 

Think of more cases, backups are crucial so changes to it should be well tested :smile:

akhilnarang committed:
refactor(restore): adjust downgrade check

First actually check whether its a downgrade, then check for the force flag

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: simplify version detection logic

Our version detection code is relatively simple, so we shouldn't have any exceptions arising there
Just check for a None return value to decide whether we should use the older logic

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor: use a function with context manager for backup decryption

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: handle older gzipped backups as well

Also fix the actual comparison

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: allow creating a backup with the older metadata style

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: add in some tests for backup

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore(installer): fix output when prompting user about downgrade

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: use source before changing its contents

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(postgres): make use of common helper

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: don't pipe output through `pv`

No point if we're using `execute_in_shell()`

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: simplify parsing version for older backups

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

vorasmit commented:
`bench restore` is failing in develop branch.
Not sure if it's just for me though.

Works well on v15-hotfix.

```
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/snv/develop/apps/frappe/frappe/utils/bench_helper.py", line 114, in <module>
    main()
  File "/home/snv/develop/apps/frappe/frappe/utils/bench_helper.py", line 20, in main
    click.Group(commands=commands)(prog_name="bench")
  File "/home/snv/develop/env/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/snv/develop/env/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/snv/develop/env/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/snv/develop/env/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/snv/develop/env/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/snv/develop/env/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/snv/develop/env/lib/python3.11/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/snv/develop/apps/frappe/frappe/commands/__init__.py", line 29, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/snv/develop/apps/frappe/frappe/commands/site.py", line 151, in restore
    _restore(
  File "/home/snv/develop/apps/frappe/frappe/commands/site.py", line 213, in _restore
    restore_backup(
  File "/home/snv/develop/apps/frappe/frappe/commands/site.py", line 265, in restore_backup
    if is_partial(sql_file_path):
       ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/snv/develop/apps/frappe/frappe/installer.py", line 803, in is_partial
    header = get_db_dump_header(sql_file_path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/snv/develop/apps/frappe/frappe/installer.py", line 873, in get_db_dump_header
    with gzip.open(file_path, "rb") as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/gzip.py", line 58, in open
    binary_file = GzipFile(filename, gz_mode, compresslevel)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/gzip.py", line 174, in __init__
    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
    
FileNotFoundError: [Errno 2] No such file or directory: '20240102_220010-****database.sql.gz'
```

akhilnarang commented:
> `bench restore` is failing in develop branch. Not sure if it's just for me though.

@vorasmit is that the whole output?

vorasmit commented:
> is that the whole output?

I missed the last line. Updated in the comment above

ankush commented:
@vorasmit can you share command you issued?

vorasmit commented:
`bench restore 20240102_220010-****-database.sql.gz`

The file is in the bench directory.
